pkg/reconciler/common/prune.go

```diff
@@ -27,527 +27,775 @@ import (
 	tknImage        string
 	targetNamespace string
-	ownerRef        v1.OwnerReference
+	ownerRef        metav1.OwnerReference
+	logger          *zap.SugaredLogger
+	ctx             context.Context
 }
 
-type pruneConfigPerNS struct {
-	config v1alpha1.Prune
+type pruneConfig struct {
+	Schedule                  string
+	Namespace                 string
+	Keep                      *uint
+	KeepSince                 *uint
+	Resources                 []string
+	ApplyToIndividualResource bool
+	TknImage                  string
 }
 
-// all the namespaces of default and the annotationbased
-type pruningNs struct {
-	commonScheduleNs map[string]*pruneConfigPerNS
-	uniqueScheduleNS map[string]*pruneConfigPerNS
-	configChanged    bool
+func Prune(ctx context.Context, k kubernetes.Interface, tektonConfig *v1alpha1.TektonConfig) error {
+	pruner, err := getPruner(ctx, k, tektonConfig)
+	if err != nil {
+		return err
+	}
+
+	return pruner.reconcile()
 }
 
-func Prune(ctx context.Context, k kubernetes.Interface, tC *v1alpha1.TektonConfig) error {
+func getPruner(ctx context.Context, k kubernetes.Interface, tektonConfig *v1alpha1.TektonConfig) (*Pruner, error) {
 	pruner := &Pruner{
-		kc:              k,
-		targetNamespace: tC.Spec.TargetNamespace,
-		ownerRef:        *v1.NewControllerRef(tC, tC.GetGroupVersionKind()),
+		tektonConfig:    tektonConfig,
+		kubeClientset:   k,
+		targetNamespace: tektonConfig.Spec.TargetNamespace,
+		ownerRef:        *metav1.NewControllerRef(tektonConfig, tektonConfig.GetGroupVersionKind()),
+		ctx:             ctx,
+		logger:          logging.FromContext(ctx),
+	}
+	return pruner, nil
+}
+
+func (pr *Pruner) reconcile() error {
+	// if an image specified in the tektonConfig spec, use that
+	if pr.tektonConfig.Spec.Pruner.Image != "" {
+		pr.tknImage = pr.tektonConfig.Spec.Pruner.Image
+	} else {
+		// get tkn cli container image name from environment
+		tknImageFromEnv := os.Getenv(prunerContainerImageEnvKey)
+		if tknImageFromEnv == "" {
+			return fmt.Errorf("tkn image '%s' environment variable not set", prunerContainerImageEnvKey)
+		}
+		pr.tknImage = tknImageFromEnv
 	}
 
-	//to remove cronjob created by older version of operator.
-	oldCronName := CronName[7:]
-	if err := pruner.checkAndDeleteCron(ctx, oldCronName); err != nil {
+	// reconcile service account used by pruner container
+	err := pr.reconcileServiceAccount()
+	if err != nil {
 		return err
 	}
 
-	//may be reconciled by removing pruning spec from tektonConfig
-	//in this case delete all the cronjobs with the label pruneCronLabel
-	if pruner.removedFromTektonConfig(tC.Spec.Pruner) {
-		cronJobs, err := pruner.listCronJobs(ctx, tC.Spec.TargetNamespace, pruneCronLabel)
-		if err != nil {
-			return err
-		}
-		if len(cronJobs.Items) > 0 {
-			for _, cronJob := range cronJobs.Items {
-				if err := pruner.deleteCronJob(ctx, cronJob.Name); err != nil {
-					return err
-				}
-			}
-		}
-		return nil
+	// reconcile role used in service account
```

Comment:
instead of creating reconcile for these three resources - serviceaccount, role binding and clusterrolebinding, we can create an installerset for them

Reply:
@piyush-garg Currently I am verifying the RBAC and SA existence and owned by the operator, if not creating new entries. In near feature(after this PR merged) I am going to add a informer to react with RBAC and SA changes (deleted/modified by external user/client).
I am afraid that adding installerset may introduce bit complication to maintain prune feature. Prune is a portable feature and all the code in one location. adding installerset may introduce to work on different places and may add bit more complication on maintenance. 

WDYT?

Reply:
@piyush-garg I have re-implemented the RBAC/SA part. Now RBAC and ServiceAccount are managed through `TektonInstallerSet`